### PR TITLE
Allow launchApp with permissions on Android

### PR DESCRIPTION
**Pull request by amanjeetsingh150**
Allow launching android apps with permission id's or as a group

For example, allowing read/write storage could be:

```
- launchApp:
     permissions: storage
```

and for user permissions or which are not in list:

```
- launchApp:
    permissions:
      android.permission.BODY_SENSORS_BACKGROUND: allow
```

### CODE REVIEW

File: maestro-client/src/main/java/maestro/drivers/AndroidDriver.kt [hunk 1]
```diff
@@ -519,26 +519,60 @@ class AndroidDriver(
...
+            permissionsResult.getOrNull()?.let {
+                it.forEach { permission ->
+                    setPermissionSafely(appId, permission, permissionValue)
+                }
+            }
+        }
     }
 
-    private fun setPermission(appId: String, permission: String, value: String) {
-        val name = permission.replace("[^A-Za-z0-9._]+".toRegex(), "")
```
Comment (berikv) [inline maestro-client/src/main/java/maestro/drivers/AndroidDriver.kt:538]: Please keep this

Comment (amanjeetsingh150) [reply maestro-client/src/main/java/maestro/drivers/AndroidDriver.kt:538]: Why was this needed btw?

Comment (berikv) [reply maestro-client/src/main/java/maestro/drivers/AndroidDriver.kt:538]: See in the chat

File: maestro-client/src/main/java/maestro/drivers/AndroidDriver.kt [hunk 2]
```diff
@@ -519,26 +519,60 @@ class AndroidDriver(
...
+                }
+                "storage" -> {
+                    setPermissionSafely(appId, "android.permission.WRITE_EXTERNAL_STORAGE", translatedPermissionValue)
+                    setPermissionSafely(appId, "android.permission.READ_EXTERNAL_STORAGE", translatedPermissionValue)
+                }
+                "calendar" -> {
+                    setPermissionSafely(appId, "android.permission.WRITE_CALENDAR", translatedPermissionValue)
+                    setPermissionSafely(appId, "android.permission.READ_CALENDAR", translatedPermissionValue)
+                }
+                else -> setPermissionSafely(appId, permission, translatedPermissionValue)
```
[the remark this answers is not in the document]
Comment (amanjeetsingh150) [reply maestro-client/src/main/java/maestro/drivers/AndroidDriver.kt:551]: Sorry but for which line 🤔

[the remark this answers is not in the document]
Comment (berikv) [reply maestro-client/src/main/java/maestro/drivers/AndroidDriver.kt:551]: Also: this implementation is not deterministic when specifying both "all" and other permissions.
The "all" permission should always be overridden by any other permissions specified.

[the remark this answers is not in the document]
Comment (berikv) [reply maestro-client/src/main/java/maestro/drivers/AndroidDriver.kt:551]: > Sorry but for which line

All lines that call `setPermissionSafely`

[the remark this answers is not in the document]
Comment (amanjeetsingh150) [reply maestro-client/src/main/java/maestro/drivers/AndroidDriver.kt:551]: Do you mean this example:

```
-launchApp:
   - permissions: { camera: deny, all: allow }
```

[the remark this answers is not in the document]
Comment (berikv) [reply maestro-client/src/main/java/maestro/drivers/AndroidDriver.kt:551]:
```
val mutable = permissions.toMutableMap()
         mutable.remove("all")?.let { value ->
             setAllPermissions(appId, value)
         }

         mutable.forEach { permission ->
             setPermission(appId, translatePermissionName(permission.key),  translatePermissionValue(permission.value))
         }

```

File: maestro-client/src/main/java/maestro/drivers/AndroidDriver.kt [hunk 3]
```diff
@@ -519,26 +519,60 @@ class AndroidDriver(
...
-        mutable.remove("all")?.let { value ->
-            setAllPermissions(appId, value)
-        }
-
-        mutable.forEach { permission ->
-            setPermission(appId, permission.key, permission.value)
+        permissions.forEach { (permission, permissionValue) ->
+            val translatedPermissionValue = translatePermissionValue(permissionValue)
+            when(permission.replace("[^A-Za-z0-9._]+".toRegex(), "")) {
+                "all" -> setAllPermissions(appId, translatedPermissionValue)
```
Comment (berikv) [inline maestro-client/src/main/java/maestro/drivers/AndroidDriver.kt:525]: This isn't deterministic. The permissions are a hash and can be in any order. When a permission is set like:

```
- launchApp: { permissions: { all: allow, notifications: deny } }
```

It depends on the runtime ordering of the hash elements weather notifications will be set to allow or deny.
Please consider reverting to what was there:

```
val mutable = permissions.toMutableMap()
         mutable.remove("all")?.let { value ->
             setAllPermissions(appId, value)
         }

         mutable.forEach { permission ->
             setPermission(appId, translatePermissionName(permission.key),  translatePermissionValue(permission.value))
         }
```

**Review by berikv — APPROVED** [review]

File: maestro-orchestra/src/main/java/maestro/orchestra/Orchestra.kt [hunk 4]
```diff
@@ -535,14 +535,15 @@ class Orchestra(
...
-            }
 
             // For testing convenience, default to allow all on app launch
             val permissions = command.permissions ?: mapOf("all" to "allow")
             maestro.setPermissions(command.appId, permissions)
 
+            if (command.clearState == true) {
+                maestro.clearAppState(command.appId)
+            }
+
```
[the remark this answers is not in the document]
Comment (amanjeetsingh150) [reply maestro-orchestra/src/main/java/maestro/orchestra/Orchestra.kt]: Done

**Pull request merged**